### PR TITLE
Implement numba overload for POTRF, LAPACK cholesky routine

### PR DESCRIPTION
**Pull request by jessegrabowski**
## Description
Tests are failing on #577 for the numba overloaded cholesky routine and I realized we didn't have a real one, so here it is. Now fully compatible with all the scipy keyword arguments. In addition, it will no longer drop numba into object mode. 

## Related Issue

- [ ] Closes #
- [ ] Related to #577

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: tests/link/numba/test_slinalg.py [hunk 1]
```diff
@@ -102,3 +107,54 @@ def test_solve_triangular_raises_on_nan_inf(value):
...
+        (
+            set_test_value(
+                pt.dmatrix(),
+                (lambda x: x.T.dot(x))(rng.random(size=(3, 3)).astype("float64")),
+            ),
+            False,
+            UserWarning,
+        ),
+    ],
+)
```
Comment (lucianopaz) [inline tests/link/numba/test_slinalg.py:142]: Since you've done quite a lot to get LAPACK's `xpotrf`multiple dtypes, don't you want to test on complex and float32 also?

Comment (jessegrabowski) [reply tests/link/numba/test_slinalg.py:142]: Yes, and complex types too. I just lazily copied the old test without updating it. Thanks for flagging this, I'll refactor the test.

Comment (jessegrabowski) [reply tests/link/numba/test_slinalg.py]: I refactored the tests so `float32` is covered. Complex is going to be a bit more complex (heh), because the ctype pointer hack I'm using to wrap the LAPACK calls doesn't work for complex inputs anymore. I remember hitting this problem in my own work a few months back, but I blocked it out of my mind. 

It's an edge case of an edge case so I'm not too worried about it, but I would prefer 100% coverage. It might mean actually figuring out how scipy binds the LAPACK functions to C code...

Comment (ricardoV94) [reply tests/link/numba/test_slinalg.py]: Feel free to raise NotImplementedError for now

File: pytensor/link/numba/dispatch/slinalg.py [hunk 2]
```diff
@@ -267,14 +292,79 @@ def solve_triangular(a, b):
...
+        return A_copy
+
+    return impl
+
+
+@numba_funcify.register(Cholesky)
+def numba_funcify_Cholesky(op, node, **kwargs):
+    lower = op.lower
+    overwrite_a = False
+    check_finite = op.on_error == "raise"
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/slinalg.py:352]: I think this flag was created because Scipy Cholesky always raised, and it was needed to catch it for some use cases. Like even if used in a switch statement where we would ignore the values it would crash the program.

That to say that raise was not necessarily a feature, but the default behavior. The flag feature was to allow suppressing the warning. So we may not need to add the extra flag logic here?

I could be thinking about the flag of another Op, but git history should confirm.

Comment (jessegrabowski) [reply pytensor/link/numba/dispatch/slinalg.py:352]: Well there is a `check_finite` flag on the original scipy function. I was going to point out that the signatures didn't match and ask if it was appropriate to change it in this PR, but nothing gets past you.

Comment (ricardoV94) [reply pytensor/link/numba/dispatch/slinalg.py:352]: What does the final suggestion look like?

Comment (jessegrabowski) [reply pytensor/link/numba/dispatch/slinalg.py:352]: Change the signature on the Cholesky `Op` to use `check_finite` instead of `on_error`, which would take a bool instead of a string. But that would break existing code, so I'd have to add a deprecation warning too.

Comment (ricardoV94) [reply pytensor/link/numba/dispatch/slinalg.py:352]: I don't think check finite has the same goal. That's just a check the input has valid values.

The op error thing was about not raising an error when the decomposition fails. I assume this can happen even with finite values?

Comment (ricardoV94) [reply pytensor/link/numba/dispatch/slinalg.py:352]: You can add the check_finite kwarg though.

About the raise, what would happen now if the matrix is not SPD?

Comment (jessegrabowski) [reply pytensor/link/numba/dispatch/slinalg.py:352]: Oh I didn't think about the difference between input and output checking. The way I'm doing the flag is currently wrong -- I'm checking the output, not the input. 

Right now if the matrix isn't PSD it will return the input matrix without changes, which is clearly wrong. I'll fix that. So the `on_error` flag should return a matrix of all nan instead of raising a "not PSD" error?

File: pytensor/tensor/slinalg.py [hunk 3]
```diff
@@ -129,8 +132,10 @@ def conjugate_solve_triangular(outer, inner):
             return [grad]
 
 
-def cholesky(x, lower=True, on_error="raise"):
-    return Blockwise(Cholesky(lower=lower, on_error=on_error))(x)
+def cholesky(x, lower=True, on_error="raise", check_finite=True):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:135]: What about default check_finite=False?

Deviates from Scipy, but avoids iterating over the array before every call. Since this wasn't a problem until now and JAX probably just drops the flag anyway, it doesn't seem too dangerous?

It also doesn't sound like it will lead to any silent bad outputs.

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:135]: Works for me

File: pytensor/link/numba/dispatch/slinalg.py [hunk 4]
```diff
@@ -267,14 +292,97 @@ def solve_triangular(a, b):
...
+    )
+
+
+@overload(_cholesky)
+def cholesky_impl(A, lower=0, overwrite_a=False, check_finite=True):
+    ensure_lapack()
+    _check_scipy_linalg_matrix(A, "cholesky")
+    dtype = A.dtype
+    if str(dtype).startswith("complex"):
+        raise ValueError(
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/slinalg.py:315]:
```suggestion
        raise NotImplementedError(
```

File: pytensor/link/numba/dispatch/slinalg.py [hunk 5]
```diff
@@ -267,14 +292,97 @@ def solve_triangular(a, b):
...
+            INFO,
+        )
+
+        return A_copy, int_ptr_to_val(INFO)
+
+    return impl
+
+
+@numba_funcify.register(Cholesky)
+def numba_funcify_Cholesky(op, node, **kwargs):
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/slinalg.py:353]: Noob question: why do we have two implementations that look almost the same? This and the numba @overloads one?

File: pytensor/link/numba/dispatch/slinalg.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/link/numba/dispatch/slinalg.py:315]: I would put this in the dispatch function. We usually assume dtypes are correct from the graph info. This will check in every call?

File: pytensor/link/numba/dispatch/slinalg.py (hunk 5 above)
Comment (ricardoV94) [reply pytensor/link/numba/dispatch/slinalg.py:353]: Also aren't we doing the finite check twice?

Comment (jessegrabowski) [reply pytensor/link/numba/dispatch/slinalg.py:353]: The overloads one is the one that tells numba how to directly call the underlying lapack function. It looks similar because in this case the boilerplace to make it happen isn't so bad. For a horrible example see [here](https://github.com/jessegrabowski/gEconpy/blob/cd2ae75747b0aa8bc23a6b35033f3a3e4ca7bc79/gEconpy/numba_tools/overloads.py#L265).

Comment (jessegrabowski) [reply pytensor/link/numba/dispatch/slinalg.py:353]: And yes you're right about the duplicated check

**Comment by ricardoV94** [comment]
Shall I squash or do you want to tidy up the git history?

**Comment by jessegrabowski** [comment]
I think it can squash? I was trying to keep thing tidy as I went. Well actually I tried to rearrange things to be ever better but it made a huge mess and I had to nuke and restart my local branch. So let me know if you want more done, and I'll do it. I'm happy on my end.
